control-plane/typeSpecConverter.cpp

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <map>
+#include <string>
+
+#include "p4/p4types.pb.h"
+
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+#include "ir/visitor.h"
+#include "lib/error.h"
+#include "lib/exceptions.h"
+#include "lib/null.h"
+
+#include "typeSpecConverter.h"
+
+namespace P4 {
+
+namespace ControlPlaneAPI {
+
+TypeSpecConverter::TypeSpecConverter(
+    P4::TypeMap* typeMap, P4::ReferenceMap* refMap, p4::P4TypeInfo* p4RtTypeInfo)
+    : typeMap(typeMap), refMap(refMap), p4RtTypeInfo(p4RtTypeInfo) {
+    CHECK_NULL(typeMap);
+    CHECK_NULL(refMap);
+}
+
+bool TypeSpecConverter::preorder(const IR::Type* type) {
+    ::error("Unexpected type %1%", type);
+    map.emplace(type, new p4::P4DataTypeSpec());
+    return false;
+}
+
+bool TypeSpecConverter::preorder(const IR::Type_Bits* type) {
+    auto typeSpec = new p4::P4DataTypeSpec();
+    auto bitTypeSpec = typeSpec->mutable_bitstring();
+    auto bw = type->width_bits();
+    if (type->isSigned) bitTypeSpec->mutable_int_()->set_bitwidth(bw);
+    else bitTypeSpec->mutable_bit()->set_bitwidth(bw);
+    map.emplace(type, typeSpec);
+    return false;
+}
+
+bool TypeSpecConverter::preorder(const IR::Type_Varbits* type) {
+    auto typeSpec = new p4::P4DataTypeSpec();
+    auto bitTypeSpec = typeSpec->mutable_bitstring();
+    bitTypeSpec->mutable_varbit()->set_max_bitwidth(type->size);
+    map.emplace(type, typeSpec);
+    return false;
+}
+
+bool TypeSpecConverter::preorder(const IR::Type_Boolean* type) {
+    auto typeSpec = new p4::P4DataTypeSpec();
+    // enable "bool" field in P4DataTypeSpec's type_spec oneof
+    (void)typeSpec->mutable_bool_();
+    map.emplace(type, typeSpec);
+    return false;
+}
+
+bool TypeSpecConverter::preorder(const IR::Type_Name* type) {
+    auto typeSpec = new p4::P4DataTypeSpec();
+    auto decl = refMap->getDeclaration(type->path, true);
+    auto name = decl->controlPlaneName();
+    if (decl->is<IR::Type_Struct>()) {
+        typeSpec->mutable_struct_()->set_name(name);
+    } else if (decl->is<IR::Type_Header>()) {
+        typeSpec->mutable_header()->set_name(name);
+    } else if (decl->is<IR::Type_HeaderUnion>()) {
+        typeSpec->mutable_header_union()->set_name(name);
+    } else if (decl->is<IR::Type_Enum>()) {
+        typeSpec->mutable_enum_()->set_name(name);
```

Comment:
Nit: why sometimes you suffix the method name with `_` and sometimes you do not?

Reply:
This is protobuf-generated code. When you use a field name in the protobuf definition that happens to be a reserved keyword in the language you are generating to, they append a `_`.